File: ruby/spec/04_trees_and_graphs/05_valid_bst.rb
require_relative './02_minimal_tree'

# n - count of nodes
# Time - O(n), Mem = O(1)
def valid_bst_a?(tree)
  valid_bst_node?(tree.root)
end

def valid_bst_node?(node)
  if left = node.left
    if left.value >= node.value || !valid_bst_node?(left)
      return false
    end
  end

  if right = node.right
    if right.value <= node.value || !valid_bst_node?(right)
      return false
    end
  end

  return true
end

RSpec.describe 'valid_bst' do
  %i(a).each do |implementation|
    describe "#{implementation} case" do
      subject { send("valid_bst_#{implementation}?", tree) }

      context 'BST tree' do
        let(:tree) { array_to_bst_a([1, 2, 3, 4, 5, 6, 7]) }

        it { is_expected.to eq true }
      end

      context 'not BST tree' do
        let(:tree) { array_to_bst_a([1, 2, 3, 4, 5, 6, 7]) }

        before do
          tree.root.left.left.value = 5
        end

        it { is_expected.to eq false }
      end
    end
  end
end

File: go/04_trees_and_graphs/05_valid_bst.go
package trees_and_graphs

import (
	. "github.com/holyketzer/ctci_v6/lib"
)

// n - count of nodes
// Time - O(n), Mem = O(1)
func ValidBstA(tree *BinaryTree) bool {
	return ValidBstNode(tree.Root)
}

func ValidBstNode(node *BTNode) bool {
	left := node.Left

	if left != nil {
		if left.Value >= node.Value || ValidBstNode(left) == false {
			return false
		}
	}

	right := node.Right
	if right != nil {
		if right.Value <= node.Value || ValidBstNode(right) == false {
			return false
		}
	}

	return true
}
